File: src/main/java/smartbrew/domain/TastingNote.java
package smartbrew.domain;

import jakarta.persistence.*;
import lombok.Data;

import java.math.BigDecimal;
import java.sql.Timestamp;

@Data
@Entity
@Table(name = "tasting_note")
public class TastingNote {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "tasting_note_id")
    private Long tastingNoteId;

    @Column(name = "fermentation_aroma", nullable = false, precision = 4, scale = 2)
    private BigDecimal fermentationAroma;

    @Column(name = "foam", nullable = false, precision = 4, scale = 2)
    private BigDecimal foam;

    @Column(name = "mouthfeel", nullable = false, precision = 4, scale = 2)
    private BigDecimal mouthfeel;

    @Column(name = "balance", nullable = false, precision = 4, scale = 2)
    private BigDecimal balance;

    @Column(name = "clarity", nullable = false, precision = 4, scale = 2)
    private BigDecimal clarity;

    @Column(name = "sweetness", nullable = false, precision = 4, scale = 2)
    private BigDecimal sweetness;

    @Column(name = "acidity", nullable = false, precision = 4, scale = 2)
    private BigDecimal acidity;

    @Column(name = "body", nullable = false, precision = 4, scale = 2)
    private BigDecimal body;

    @Column(name = "grainy_aroma", nullable = false, precision = 4, scale = 2)
    private BigDecimal grainyAroma;

    @Column(name = "finish", precision = 4, scale = 2)
    private BigDecimal finish;

    @Column(name = "evaluator_name", nullable = false, length = 10)
    private String evaluatorName;

    @ManyToOne
    @JoinColumn(name = "batch_id")
    private Batch batch;

    @Column(name = "created_at", nullable = false)
    private Timestamp createdAt = new Timestamp(System.currentTimeMillis());

    @Column(name = "updated_at")
    private Timestamp updatedAt;
}

File: models/tasting_note.js
module.exports = (sequelize, DataTypes) => {
    const TastingNote = sequelize.define(
        'TastingNote',
        {
            tasting_note_id: {
                type: DataTypes.INTEGER,
                primaryKey: true,
                autoIncrement: true,
                allowNull: false,
            },
            fermentation_aroma: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            foam: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            mouthfeel: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            balance: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            clarity: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            sweetness: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            acidity: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            body: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            grainy_aroma: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: false,
            },
            finish: {
                type: DataTypes.DECIMAL(4, 2),
                allowNull: true,
            },
            evaluator_name: {
                type: DataTypes.STRING(10),
                allowNull: false,
            },
            batch_id: {
                type: DataTypes.INTEGER,
                references: { model: 'Batch', key: 'batch_id' },
                onDelete: 'SET NULL',
                onUpdate: 'CASCADE',
            },
            created_at: {
                type: DataTypes.DATE,
                defaultValue: DataTypes.NOW,
            },
            updated_at: {
                type: DataTypes.DATE,
                allowNull: true,
            },
        },
        {
            tableName: 'tasting_note',
            timestamps: false,
        },
    );

    TastingNote.associate = (models) => {
        TastingNote.belongsTo(models.Batch, { foreignKey: 'batch_id' });
    };

    return TastingNote;
};
